Add stats endpoint to GeoIpDownloader

probakowski opened the pull request:
This change adds `_geoip/stats` endpoint that can be used to collect basic data about geoip downloader (successful, failed and skipped downloads, current db count and total time spent downloading).
It also fixes missing/wrong origins for clients that will break if used with security.

Relates to https://github.com/elastic/elasticsearch/issues/68920

probakowski committed:
Add stats endpoint to GeoIpDownloader

original-brownbear committed:
Fix testRetentionWithMultipleRepositories Failure (#70202)

The wait for green check times out in this test because the datastream
that is backing the SLM history store is running into a 60s delayed allocation
when the test stops a data node that holds some of its shards.
Fixed by turning off the delayed allocation in this test.
It appears this only affects this one test because it's the only test that
creates a policy and executes it before stopping a node

closes #70185

iverase committed:
Fix overflow in GeoTileGridTiler (#70222)

This PR uses longs to compute the number of tiles touching a bounding box

martijnvg commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/RestGeoIpDownloaderStatsAction.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) {
+        ListTasksRequest listTasksRequest = new ListTasksRequest()
```
I think should be done in a transport action (can be as simple as `HandledTransportAction`), the rest layer should just be a layer that translates a rest request to an action request and delegates to transport action.

This can make applying more fine grained security privileges possible (otherwise we always need to grant privileges on the list task action, which exposes a lot of other information) and this 
approach we do for all apis/actions, even if these actions are very simple.

probakowski replied:
I've added relevant actions

martijnvg reviewed (commented):
I think the information reported in `GeoIpDownloaderStats` looks good.

However I think it be also great if this api also returns information about each `DatabaseRegistry` running on each ingest node. For example report about the available lazy loaders and perhaps report what files are in the geoip tmp directories. I think this can make debugging easier. Wdyt?

If we would do this then this does mean that the transport action would need to extend something `TransportNodesAction` class (which would help sending an internal request to each ingest node). And as part of this api we would also make a call to list tasks action in order to get stats about the geoip downloader.

Commit:
[Transform] deprecate "transform.node" node attribute (#70234)

The node attribute transform.node has been introduced in #52712 (7.7) to implement dedicated
transform nodes. A technical limitation in the stack made it necessary to use this workaround.
This limitation has been removed by #54998 (7.9). It is now possible to find out, if a node is a
transform node without using the node attribute. This PR deprecates the node attribute and bases
the placement decision on the node role. The attribute gets unused for placement, however we can
not remove the node attribute yet, because in a mixed version cluster the attribute might still
be used, especially when master gets updated last. The node attribute will be completely removed
in 8, as the upgrade path requires to upgrade to 7.last before upgrading to 8.x.

andreidan committed:
Add unit test for shard bulk stats reporting (#70254)

ywelsch committed:
Track query stats bug (#70273)

The query cache stats can currently report negative memory usage, which breaks REST responses for indices stats, node stats, etc, see #55434 as one such example.

The reason why negative memory usage is reported is because of the following trappy calculation:

https://github.com/elastic/elasticsearch/blob/1de0b616ebb92de6060510e92269ef87fc6a8649/server/src/main/java/org/elasticsearch/indices/IndicesQueryCache.java#L98-L101

- weight can be `Double.POSITIVE_INFINITY` when `totalSize == 0` and `stats.size() == 0`
- if we then also have `sharedRamBytesUsed > 0`, then `additionalRamBytesUsed` will be `Long.MAX_VALUE`
- if you then sum up multiple of these `Long.MAX_VALUE` values (one for each shard), this leads to an overflow, resulting in negative numbers as seen in #55434.

The reason sharedRamBytesUsed can be > 0 when there are no shard stats at all is because of the memory occupied by `LRUQueryCache.uniqueQueries`, where the lifetime can extend even closing of shards (where they are removed from shardStats).

Note that a workaround to the above bug is to [clear the cache](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-clearcache.html), as it makes `sharedRamBytesUsed == 0`, see https://github.com/elastic/elasticsearch/blob/1de0b616ebb92de6060510e92269ef87fc6a8649/server/src/main/java/org/elasticsearch/indices/IndicesQueryCache.java#L168-L174

Closes #55434

csoulios committed:
[RollupV2] Move rollup metadata to field mappings (#69921)

This PR moves field related rollup metadata from the index metadata to the field mapping metadata.

- date_histogram fields is moved to the timestamp field 
     (fields are fixed_interval or calendar_interval, time_zone)
- histogram fields are moved to the numeric field on which the the histogram is computed 
     (the field is named interval)

Also, the index uuid has been added to the index rollup settings as index.rollup.source.uuid 
and index.rollup.source.name

The rest of the RollupMetadata has been removed and no rollup metadata exists in the global 
cluster state.

Co-authored-by: Tal Levy <jubboy333@gmail.com>

martijnvg committed:
Support specifying multiple templates names in delete composable index template api (#70094)

Add support to delete composable index templates api to specify multiple template
names separated by a comma.

Change to cleanup template logic for rest tests to remove all composable index templates via a single delete composable index template request. This to optimize the cleanup logic. After each rest test we delete all templates. So deleting templates this via a single api call (and thus single cluster state update) saves a lot of time considering the number of rest tests.

If this pr is accepted then I will do the same change for the delete component template api.

Relates to #69973

andreidan committed:
Fix rollover alias definition in templates validation (#70259)

This fixes the case where a valid composable template overrides/shadows a
legacy template that defines the rollover alias. This is a valid configuration
and rollover should not fail.

martijnvg committed:
disable bwc tests for backporting #70094

Commit:
Completely disallow setting negative size in search (#70209)

We used to treat setting size to -1 in search request bodies or as a rest
parameter as a no-op, using the default search size of 10 in this case. This
lenient behaviour was deprecated in #69548 and is removed with this PR in 8.0.

Relates to #69548

probakowski committed:
add nodes info, add action/transport action

probakowski committed:
Merge remote-tracking branch 'origin/master' into geoip-stats

probakowski committed:
imports

probakowski committed:
fix executor initialization

martijnvg commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/stats/GeoIpDownloaderStatsAction.java:
```diff
@@ -0,0 +1,159 @@
+                }
+                builder.startObject(e.getKey());
+                if (response.databases.isEmpty() == false) {
+                    builder.array("databases", response.databases.toArray(String[]::new));
```
Perhaps return an array of objects here? This leaves room for adding more properties about the databases in the future without breaking bwc. For example like database type and its current usages.

probakowski replied:
good idea, I'll change that and add tests

probakowski replied:
done

martijnvg reviewed (commented):
👍 I think this looks good. Let's add tests and turn this into a regular pr?

probakowski committed:
tests

probakowski committed:
tests

probakowski committed:
Merge remote-tracking branch 'origin/master' into geoip-stats

probakowski committed:
tests

probakowski committed:
add HLRC endpoints

probakowski committed:
rename + tests

probakowski committed:
tests

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

probakowski committed:
build fix?

probakowski committed:
build fix?

probakowski committed:
test fix

probakowski committed:
fix test

probakowski committed:
test fix

probakowski committed:
imports

probakowski committed:
test

probakowski commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into geoip-stats

probakowski commented:
@elasticmachine run elasticsearch-ci/2

probakowski commented:
@elasticmachine run elasticsearch-ci/2

martijnvg commented on modules/ingest-geoip/src/yamlRestTest/resources/rest-api-spec/test/ingest_geoip/30_geoip_stats.yml:
```diff
@@ -0,0 +1,10 @@
+"Test geoip stats":
+  - do:
+      ingest.geo_ip_stats: {}
+  - match: { stats.successful_downloads: 0 }
```
I think this may fail if during the yaml tests execution the geoip downloader actually downloads databases?
Perhaps we should just validate the existence of specific fields (using `is_true`)?

probakowski replied:
it shouldn't downloader is disabled for these tests but it's perhaps safer to use `is_true` as you suggested, I'll update it

probakowski replied:
I used `gte` - `is_true` doesn't work with 0 ( it treats it as `false`)

martijnvg replied:
:( too bad that there isn't another way to check the existence of a field.

martijnvg reviewed (approved):
LGTM 👍

probakowski committed:
make yaml test to be more permisive

probakowski commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into geoip-stats
